### PR TITLE
add option to (left) join collection while reading nodes

### PR DESCRIPTION
**Pull request by abred**
Example use case:
For each node in nodes collection we want to store scores for some property.
The scores are computed by some model. We can have different versions of that
model. Instead of storing each version as attributes in the nodes collection
using different names we store them as separate collections.
When loading the nodes we (left) join the chosen collection onto the nodes
using the node ID.

### CODE REVIEW

**Comment by trivoldus28** [comment]
@abred If I understand this feature correctly, let's say for `join_collection` = "v1", you're storing some attributes, let's say `a` and `b`, and for some other `join_collection` "v1a", the attributes could be `a` and `c`, and you'd want to retrieve just these by specifying `join_collection`?

How would the user specify these different attributes for different `join_collection`s? I would think there should be some added function for the line "join_col_entry = self.database[join_collection].find_one()" to work.

What about just specifying different attributes from the user script by using `read_attrs`?

**Comment by abred** [comment]
@trivoldus28 The use case is this:
Let's say I have a collection `nodes` with some attributes, e.g. `x,y,z,t`. For each node I have values `score_a` and `score_b`. However I have different versions of these values (for instance different models I want to compare). Now instead of "polluting" the main collection with `score_a_v1`, `score_b_v1`, `score_a_v2`, ... I have separate collections `vN`. Each collection contains for each node an entry with `score_a` and `score_b` and the matching ID. The user can then just set `join_collection` to whatever version should be used and the rest of the code can act as if the value had always been part of the nodes.

The code is agnostic to whatever attributes are in the `join_collection`, all are merged, and there can only be one collection joined.

**Comment by trivoldus28** [comment]
@abred Thanks for the explanation! Just making sure, you have the current node collection `nodes_collection` (i.e., table A) that you're querying `read_attrs` for certain nodes, and now you want to merge matching `read_attrs` values from another collection `join_collection` (i.e., table B), if they exist there? What if the attrs don't exist in table A but in table B? Is this specified with a left join?

**Comment by trivoldus28** [comment]
@abred I wonder if you could write some tests for this feature to make sure it doesn't get broken with future commits? I think we also need to put back the parked tests as much as possible but unfortunately I'm a bit occupied right now... Would you be up for it, at least with the graph tests?

**Comment by abred** [comment]
@trivoldus28 I am not sure about the exact definitions of database joins, how it behaves in edge cases etc, so I don't know if it is completely correct to call it a left join, but it is close?

The existing code already assumes that the `nodes_collection` has an attribute `id`. The new code uses that to merge matching entries in `nodes_collection` and `join_collection`. As it uses `id` the matches should be unique. It seems if there is no matching entry for an `id` in `join_collection` it is dropped completely.
Currently the projection on `read_attrs` is performed after the merge. So the two collections are merged and then only the attributes contained in `read_attrs` are kept (if `read_attrs` is set)

I pushed some changes (the projection part was not correct) and added some tests.

File: daisy/persistence/mongodb_graph_provider.py [hunk 1]
```diff
@@ -160,6 +165,11 @@ def read_nodes(self, roi, attr_filter=None, read_attrs=None):
             read_attrs (``list`` of ``string``):
 
                 Attributes to return. Others will be ignored
+
```
Comment (trivoldus28) [inline daisy/persistence/mongodb_graph_provider.py:168]: Can we have a brief description of the expected behavior and fault conditions here? I feel like "left join" is not quite an accurate definition in this case (and even if it is, would help unfamiliar users).

File: tests/test_graph_provider_join.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    def delete_db(self):
+        self.client.drop_database(self.db_name)
+
+    def mongo_provider_factory(self, mode):
+        return daisy.persistence.MongoDbGraphProvider(
+            self.db_name,
+            mode=mode,
+            position_attribute=['x', 'y'])
+
+    def test_join_collection(self):
```
Comment (trivoldus28) [inline tests/test_graph_provider_join.py:37]: Can you provide a brief high level description and expected behavior for each test to help me understand what you're testing?

**Review by trivoldus28 — CHANGES_REQUESTED** [review]
Left some comments. Seems like some tests are still failing do you know what's going on?

**Comment by abred** [comment]
The error does not happen in one of the tests I added.
Nevertheless, if you have a look a the travis output
https://github.com/funkelab/daisy/blob/620f4dd6a97439f56a2029ed407d647a3ddb0445/daisy/tcp/tcp_server.py#L80-L82
`E       RuntimeError: Set changed size during iteration`

after the stream is closed it is set to `None`. I guess in unlucky cases the garbage collector runs early, removes the object while the loop is still running and then we get this error. I did not get this error when running the tests locally.
It might be sufficient to just change it to:
```
        for stream in list(self.client_streams):
```

**Comment by trivoldus28** [comment]
Yea I think that's one currently benign bug that we'd need to look into fixing. Once you added a little more documentation I think we can merge it in.

File: daisy/persistence/mongodb_graph_provider.py [hunk 3]
```diff
@@ -160,6 +165,20 @@ def read_nodes(self, roi, attr_filter=None, read_attrs=None):
...
+            join_collection (``string``):
+
+                Compute (left) join of the nodes collection and this
+                collection using the id attribute.
+                Some node attributes can be stored in another collection.
+                Assumes that all nodes have the same attributes.
```
Comment (trivoldus28) [inline daisy/persistence/mongodb_graph_provider.py]: Potentially under defined. What do you mean by assuming that all nodes have the same attributes? I think in this case the attributes are the `read_attrs` right? It would also help to add that nodes that don't have all `read_attrs` are silently ignored. Maybe the description can be expanded with something like: "Attributes in nodes in the `join_collection` that have matching `id`s are joined to the main collection. Joined nodes that don't have all required `read_attrs` are silently dropped." Would this be accurate based on the provided tests?

**Review by trivoldus28 — CHANGES_REQUESTED** [review]

File: daisy/persistence/mongodb_graph_provider.py (hunk 3 above)
Comment (abred) [reply daisy/persistence/mongodb_graph_provider.py]: `read_attrs` is optional.
If `read_attrs` is set the output nodes can have **at most** the attributes in `read_attrs`. If there is an attribute in `read_attrs` but not in the `nodes` collection (or in the `join_collection`) that node will still be in the output, just without this attribute (unchanged behavior to before).
If `read_attrs` is not set and `join_collection` is not set we just take whatever is in each node.
However, in principle, we cannot really assume anything about the content of each node entry as in this type of database each entry could be completely different.
And that is a bit problematic if `read_attrs` is not set but `join_collection` is set. Due to the way entries are matched across collections in mongodb ("$lookup") we have to do a projection afterwards (otherwise the attributes from the `join_collection` are nested within an array/dict). And a projection can contain either negative entries (take all except these) or positive entries (take only these) but not both. So because we have to do a (positive) projection we have to list all attributes that we want in the output, but we do not know the attributes in the node entries. So what I am doing is taking a random entry and adding all the attributes in that entry to the projection. There might be extra attributes in this entry compared to other entries, that is ignored, but there also might be missing attributes in this entry and those will be discarded from the other entries, too.
I guess one (ugly) alternative could be to loop over all nodes at the end, and fix them manually, instead of the projection. 
I hope this makes sense, please ask if I should clarify anything

Comment (trivoldus28) [reply daisy/persistence/mongodb_graph_provider.py]: Thanks for the explanation! For the `read_attrs` is not set but `join_collection` is set case, can't you just look up all matching `id`s in the `join_collection` without filtering with projections and join up the available attributes in both collections?

Comment (abred) [reply daisy/persistence/mongodb_graph_provider.py]: the tasks are `lookup`, `unwind`, `projection`
`lookup` does a "left outer join".
nodes collection:
```
id:1
```
join_collection:
```
id: 1
_id: hash
scoreA: 1
scoreB: 2
```

The result of the lookup op is 
```
id: 1
joined: [{id: 1, _id: hash, scoreA: 1, scoreB: 2}]
```
The list in `joined` could contain several entries, if the id occurs several times. `unwind` would split those into separate entries. However if there is only one it just removes the list
```
id: 1
joined: {id: 1, _id: hash, scoreA: 1, scoreB: 2}
```
The projection flattens the nested dict, removes the duplicate `id` attribute and removes the `_id` attribute
```
id: 1
scoreA: 1
scoreB: 2
```

So the projection has two different tasks, depending on the arguments.
If `read_attrs` is set it should filter the attributes.
If `join_collection` is set it should flatten the nested dicts and clean up the result. However it still has the "side effect" of doing the filtering.

Comment (trivoldus28) [reply daisy/persistence/mongodb_graph_provider.py]: Regardless of implementation (i.e., the use of `lookup`, `unwind`, etc...) at the high level I was thinking that the right thing to do is to _not_ do the filtering when `join_collection` is used but not `read_attrs`, right? Just grab all attributes in the node in `nodes_collection`, then in the node with same `id` in `join_collection`, then join and return everything. Wouldn't that be correct?

For the case where both `join_collection` and `read_attrs` are used, do the same thing but set the correct projection so only the queried attrs are returned. I think you can do this over two separate queries, one for `nodes_collection` and another for `join_collection`. You can optimize it to be 1 single query later once the behavior is more defined.

Comment (abred) [reply daisy/persistence/mongodb_graph_provider.py]: @trivoldus28 I pushed on update. It now uses python list/dict comprehension to perform the join etc manually instead of pymongo functions. Likely slower (especially for larger collections) but hopefully with fewer assumptions.

Comment (trivoldus28) [reply daisy/persistence/mongodb_graph_provider.py]: @abred Ok looks good! Sorry for the delay. With this reference implementation you probably can now optimize and see how much more performance it gives you.

Just one quick question: this implementation returns a full `join`, and not just a `left join` right? I'm looking at this diagram: https://stackoverflow.com/questions/53949197/isnt-sql-a-left-join-b-just-a

Comment (abred) [reply daisy/persistence/mongodb_graph_provider.py]: I don't think I will optimize it. As far as I can see I cannot get the identical behavior with the mongodb functionality (as it is not a relational database but a document-based one there are no guarantees for the content of each document), but I think it is fine.

No, I think it is sitll a `left join`, it takes the rows in `A` and attaches the attributes from `B`, but if there is a row in `B` that is not included in `A` it will not be part of the result.

Comment (trivoldus28) [reply daisy/persistence/mongodb_graph_provider.py]: Okay, thanks for the clarification!
